include/deal.II/dofs/dof_vector.h

```diff
@@ -0,0 +1,243 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2007 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__dof_vector_h
+#define dealii__dof_vector_h
+
+#include <deal.II/base/smartpointer.h>
+#include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/vector_memory.h>
+
+DEAL_II_NAMESPACE_OPEN
+/**
+ * An object storing a vector describing a finite element function
+ * together with its DoFHandler and the ConstraintMatrix associated
+ * with the handler.
+ *
+ * Data vectors for finite element discretizations do not make sense
+ * without their DoFHandler. Nevertheless, deal.II has assumed for
+ * decades, that there is a single DoFHandler in a program and that
+ * all vectors refer to it. This class encapsulates both, to begin
+ * with in a minimally intrusive way, giving access to both
+ * structures.
+ *
+ * @todo Ensure that both objects are synchronized, that is, the
+ * DoFHandler is not changes without the vector being resized.
+ */
+template <class DH, class VECTOR=Vector<double> >
+class DoFVector :  public Subscriptor
+{
+public:
+  /// @name Constructors and initialization
+  ///@{
+  /**
+   * Construct an object with a DoFHandler and borrow a vector from
+   * GrowingVectorMemory for its data. This vector is then owned by
+   * this object and released when the object itself expires.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints);
+
+  /**
+   * Construct an object with a DoFHandler but without constraints and
+   * borrow a vector from GrowingVectorMemory for its data. This
+   * vector is then owned by this object and released when the object
+   * itself expires.
+   *
+   * @note This constructor uses the ConstraintMatrix
+   * #no_constraints. The created object thus can output vectors in
+   * distributed format, but writing into vectors will most likely
+   * fail without using a constraint matrix from somewhere else.
```

Comment:
I don't think I understand what this paragraph means:

* Is `no_constraints` an internal variable or something the user is supposed to understand?

* What does "output vectors in distributed format" mean?

* Why would writing into vectors fail? Do you want to say that doing something like copy-local-to-global will yield the *wrong results* without an appropriate constraints object?